compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalJvmCompilerRunner.kt

```diff
@@ -456,24 +470,118 @@ class IncrementalJvmCompilerRunner(
+            reporter.measure(BuildTime.SHRINK_AND_SAVE_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
                 shrinkAndSaveClasspathSnapshot(classpathChanges, caches.lookupCache)
             }
         }
     }
 
+    /** Used by [shrinkAndSaveClasspathSnapshot]. */
+    private sealed class ShrinkMode {
+        object NoChanges : ShrinkMode()
+
+        class IncrementalNoNewLookups(
+            val shrunkCurrentClasspathAgainstCurrentLookups: List<ClassSnapshotWithHash>,
+        ) : ShrinkMode()
+
+        class Incremental(
+            val currentClasspathSnapshot: List<ClassSnapshotWithHash>,
+            val shrunkCurrentClasspathAgainstPreviousLookups: List<ClassSnapshotWithHash>,
+            val addedOrUpdatedLookupSymbols: Set<LookupSymbolKey>
+        ) : ShrinkMode()
+
+        object NonIncremental : ShrinkMode()
+    }
+
     private fun shrinkAndSaveClasspathSnapshot(classpathChanges: ClasspathChanges.ClasspathSnapshotEnabled, lookupStorage: LookupStorage) {
-        val classpathSnapshot = currentClasspathSnapshot ?: reporter.measure(BuildTime.LOAD_CLASSPATH_SNAPSHOT) {
-            CachedClasspathSnapshotSerializer.load(classpathChanges.classpathSnapshotFiles.currentClasspathEntrySnapshotFiles)
+        // In the following, we'll try to shrink the classpath snapshot incrementally when possible.
+        // For incremental shrinking, we currently use only lookupStorage.addedOrUpdatedLookupSymbols, not
+        // lookupStorage.removedLookupSymbols. It is because:
+        //   1. Tracking removedLookupSymbols is expensive.
+        //   2. Updating the shrunk classpath snapshot based on removedLookupSymbols is also expensive.
+        // Without (2), the shrunk classpath snapshot might be slightly larger than necessary, but it is okay for it to be an
+        // over-approximation.
+        val shrinkMode = when (classpathChanges) {
+            is NoChanges -> {
+                val addedOrUpdatedLookupSymbols = lookupStorage.addedOrUpdatedLookupSymbols
+                if (addedOrUpdatedLookupSymbols.isEmpty()) {
+                    ShrinkMode.NoChanges
+                } else {
+                    val shrunkPreviousClasspathAgainstPreviousLookups =
+                        reporter.measure(BuildTime.LOAD_SHRUNK_PREVIOUS_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
+                            ListExternalizer(ClassSnapshotWithHashExternalizer)
+                                .loadFromFile(classpathChanges.classpathSnapshotFiles.shrunkPreviousClasspathSnapshotFile)
+                        }
+                    ShrinkMode.Incremental(
+                        currentClasspathSnapshot = reporter.measure(BuildTime.LOAD_CURRENT_CLASSPATH_SNAPSHOT_AFTER_COMPILATION) {
+                            CachedClasspathSnapshotSerializer
+                                .load(classpathChanges.classpathSnapshotFiles.currentClasspathEntrySnapshotFiles)
```

Comment:
This initialization logic is duplicated, can you make the backing property lazy?

Reply:
Hm they're in different classes now so it makes sharing harder.

Hopefully it's just 2 method calls so duplication is tolerable. (Also, if the call to `removeDuplicateAndInaccessibleClasses` is missing, the code will not compile due to incompatible types, so it's kind of a safeguard.)